### PR TITLE
Feature/data context namespace

### PR DESCRIPTION
**Pull request by jcampbell**
This PR:
  1. Bring the latest changes from develop into feature/data_context
  2. Adds a new generator GlobReaderGenerator that can accept glob patterns to build data_assets, with associated tests
  3. Refactors develop changes into the new reader API
  4. Updates documentation for new API

### CODE REVIEW

File: docs/source/data_contexts.rst [hunk 1]
```diff
@@ -22,6 +22,7 @@ There are currently four types of data contexts:
   - :ref:`PandasCSVDataContext`: The PandasCSVDataContext ('PandasCSV') exposes a local directory containing files as datasets.
   - :ref:`SqlAlchemyDataContext`: The SqlAlchemyDataContext ('SqlAlchemy') exposes tables from a SQL-compliant database as datasets.
   - :ref:`SparkCSVDataContext`: The SparkCSVDataContext ('SparkCSV') exposes csv files accessible from a SparkSQL context.
+  - :ref:`SparkParquetDataContext`: The SparkParquetDataContext ('SparkParquet') exposes parquet files accessible from a SparkSQL context.
```
Comment (eugmandel) [inline docs/source/data_contexts.rst:25]: Aren't the old *DataContext classes are going away? How should we deal with deprecation?

File: great_expectations/data_context/data_context.py [hunk 2]
```diff
@@ -39,11 +41,37 @@
...
+
+    DataContexts use a datasource-based namespace, where each accessible type of data has a four-part
+    normaled data_asset_name, consisting of:
+    
+    **datasource / generator / generator_asset / expectation_suite**
+
+      - The datasource actually connects to a source of materialized data and returns Great Expectations
+        DataAssets connected to a compute environment and ready for validation.
+
+      - The Generator knows how to intropsect datasources and produce identifying "batch_kwargs" that define
```
Comment (eugmandel) [inline great_expectations/data_context/data_context.py]: typo: intropsect -> introspect

**Comment by jcampbell** [comment]
I just added two new features into this PR:
1. Change autogenerated run_id from UUID to ISO8601 UTC timestamp at now()
2. Added preliminary support for get_validation_result (but it's brittle around run_ids)

I've also started adding content to the roadmap/changelog and will continue updating that.

File: great_expectations/data_context/data_context.py [hunk 3]
```diff
@@ -953,12 +981,11 @@ def profile_datasource(self, datasource_name, generator_name=None, profiler=Basi
...
                 logger.info("\tProfiled %d rows from %s (%.3f sec)" % (row_count, name, duration))
 
-            #!!! FIXME: THIS IS WAAAAY TO GENERAL. As soon as BatchKwargsError is fully implemented, we'll want to switch to that.
-            # TODO: ^^^
             except ProfilerError as err:
                 logger.warning(err.message)
-            except Exception as exc:
-                logger.warning("\tSomething went wrong when profiling %s. (Perhaps a loading error?) Skipping." % (name))
+            except IOError as exc:
+                logger.warning("\IOError while profiling %s. (Perhaps a loading error?) Skipping." % (name))
```
Comment (eugmandel) [inline great_expectations/data_context/data_context.py]: Are we letting all other errors propagate all the way up?

**Review by eugmandel — CHANGES_REQUESTED** [review]

File: docs/source/data_contexts.rst (hunk 1 above)
Comment (jcampbell) [reply docs/source/data_contexts.rst:25]: Yes; I think we should document it, and view that as part of the comprehensive documentation update (but I wanted to keep all the docs in sync with bringing in develop first)

File: great_expectations/data_context/data_context.py (hunk 2 above)
Comment (jcampbell) [reply great_expectations/data_context/data_context.py]: Good catch; will fix

File: great_expectations/data_context/data_context.py (hunk 3 above)
Comment (jcampbell) [reply great_expectations/data_context/data_context.py]: I'm not sure yet what other errors might be...but before the behavior was to eat *any* error, which I think of as generally bad practice so I figured we'd let them go up if such errors exist and develop a testing plan once we know what other errors might be

**Comment by abegong** [comment]
Is this mergeable?

**Pull request merged**